### DISCUSSION TITLE
[feat] add dead link checker to CI/CD

### DISCUSSION THREAD

**Discussion by Anonymitaet** [apache/pulsar#19144] [Ideas]
# Issue

Sometimes 404 errors occur in docs.

Currently, we use the following ways to check them:

### Method 1: use tools to check them manually at regular intervals

  - [Dr. Link Checker](https://www.drlinkcheck.com/)
  
  - [W3C Link Checker](https://validator.w3.org/checklink?uri=http%3A%2F%2Fpulsar.apache.org%2Fdocs&hide_type=all&recursive=on&depth=4&check=Check)
  
  - [Dead Link Checker](https://www.deadlinkchecker.com/)

#### Disadvantages

- It's easy to forget to do this task.

- It entirely depends on the contributor's willingness.

### Method 2: check the links through the local preview process

It reminds you of 404 errors if any.

#### Disadvantages

- Some contributors need to preview docs on the local machines and the link check step is skipped.

- Even though contributors preview docs and get reminders of 404 issues, they can ignore this error and still submit doc PRs.

# Solution

Can we implement a `dead link check` to the CI/CD process in `pulsar-site`?

### Example

Skywalking
<img width="1017" alt="image" src="https://user-images.githubusercontent.com/50226895/210919627-62826cd9-8998-4169-8fa8-3eba19cd73b2.png">

### Benefits

- Docs submitted from all contributors can be checked in a prominent and easy way

- If we want to ensure no 404 links in docs, we can set this check as `required`. In this way, PRs can not be merged with 404 issues.

- (Other doc-related checks, e.g., **grammar checks, typo checks**, etc can also be implemented)

cc @momo-jun @DaveDuggins @D-2-Ed

**Comment by tisonkun** [comment]
FYI: I'm going to enable Docusaurus builtin broken links checker in the site repo in this month.

* https://docusaurus.io/docs/api/docusaurus-config#onBrokenLinks
* https://docusaurus.io/docs/api/docusaurus-config#onBrokenMarkdownLinks

I'll enable the warning level first to get a report and see how to fix.

Other integrations proposed here can be good to have a try. But I think we should enable them on the site repo, right?

**Reply by Anonymitaet** [reply]
> I think we should enable them on the site repo, right?

Yes

**Comment by momo-jun** [comment]
The Docu built-in checker sounds better. Does the detection happen during the local preview?
If yes, it would be more proactive than a checker in the CI/CD process. But on the other hand, not all contributors run the local preview. It's not a mandatory step for minor fixes. The checker integrated in the CI/CD process can be a gate for this.

**Reply by tisonkun** [reply]
> Does the detection happen during the local preview

Yes. It will.
